### PR TITLE
Avalara Connector - COMMERCE-4276 + subtasks, COMMERCE-4348

### PR DESCRIPTION
**Pull request by katienesterovich**
### [COMMERCE-5371](https://issues.liferay.com/browse/COMMERCE-5371) - Move Avalara credentials configuration to a global scope

- Application Menu (Under Settings > Avalara)

image

- Currently one tab - credentials

image

- Failed Verify Connection 

image

- Successful verification

image

Updated channel Avalara settings 
- If Global credentials invalid: 
image
- If Global credentials valid: 
image

### [COMMERCE-5297](https://issues.liferay.com/browse/COMMERCE-5297) - Make Company Code input a dropdown with avalara values
- Company codes are now displayed in a dropdown (Company name is displayed, but company Code is registered)
- If not specified or get call failed, will be set to DEFAULT
   image

### [COMMERCE-5281](https://issues.liferay.com/browse/COMMERCE-5281) - Create avalara dispatch trigger
- Modeled after existing dispatch modules - only TaskExecuter needed to be implemented for dispatch
- Our helper will create/update ~42k entries every time the job is run
- We will recommend users run the regions/countries health checks prior to running the jobs, as it will result in a more complete dataset 

### [COMMERCE-5282](https://issues.liferay.com/browse/COMMERCE-5282) - Automate trigger creation and deletion
- Trigger created and marked 'active' on avalara-tax-engine activation PER CHANNEL
   - Name convention `Avalara-[ChannelID]`
- Trigger de-activated/re-activated when when CHANNEL Avalara tax engine disabled/enabled. 
- Trigger deleted on before channel deletion, along with the tax rate rels
- Currently, the cron-expression is set to run every month on the first at 0:00am
- The job will fail if global credentials are invalidated after channel activation (no model listener there)

### [COMMERCE-5296](https://issues.liferay.com/browse/COMMERCE-5296) - Add a 'Run Now' button in Avalara settings
- Added logic and error messages to facilitate activation workflow
   - If valid credentials configured
       - Settings panel is shown
       - Prompted to run the job for the first time
     
image

   - While job is running

image

   - After successful run

image

   - After a failed run

image

### [COMMERCE-5298](https://issues.liferay.com/browse/COMMERCE-5298)- Display AvalaraTaxRate entries in a new tab
- Reusing most of existing logic from by-adress, copied over and cleaned up for avalara. This is done to prevent exposing connector logic to core/dxp apps.
- User can edit/delete entries (I was not able to remove that menu)
- Search bar present but does not work (I am experiencing same issue with other tax engine search bars)
- If user has not previously run health checks but has run the sync job, taxCategoryId will be 0 and this view will error out. To fix, user will need to run health checks and manually rerun the job from the dispatch menu.
- There will be ~ 42k entries 

image

### [COMMERCE-4348](https://issues.liferay.com/browse/COMMERCE-4348) - Add a health check to map Tax Categories to Avalara TaxCodes

- Health check

image

- Tax Categories created (Expired categories are filtered out in the Avalara call)

image

### CODE REVIEW

**Comment by dsitu** [comment]
ci:test:sf

**Comment by katienesterovich** [comment]
ci:test:sf

**Comment by katienesterovich** [comment]
ci:test:relevant

**Comment by riccardo-alberti** [comment]
ci:test:sf

**Comment by katienesterovich** [comment]
ci:test:sf

**Comment by katienesterovich** [comment]
ci:test:sf

**Comment by katienesterovich** [comment]
ci:test:relevant

File: modules/dxp/apps/commerce-avalara/commerce-avalara-connector/src/main/java/com/liferay/commerce/avalara/connector/internal/helper/CommerceAvalaraDispatchTriggerHelperImpl.java [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+		return _dispatchTriggerLocalService.fetchDispatchTrigger(
+			commerceTaxMethod.getCompanyId(), triggerName);
+	}
+
+	private DynamicQuery _getDispatchTriggerByIdQuery(
+		DispatchTrigger dispatchTrigger) {
+
+		DynamicQuery dynamicQuery = _dispatchLogLocalService.dynamicQuery();
+		Property dispatchTriggerId = PropertyFactoryUtil.forName(
+			"dispatchTriggerId");
```
Comment (dsitu) [inline modules/dxp/apps/commerce-avalara/commerce-avalara-connector/src/main/java/com/liferay/commerce/avalara/connector/internal/helper/CommerceAvalaraDispatchTriggerHelperImpl.java:230]: new line here
```suggestion
		DynamicQuery dynamicQuery = _dispatchLogLocalService.dynamicQuery();

		Property dispatchTriggerId = PropertyFactoryUtil.forName(
			"dispatchTriggerId");
```

File: modules/dxp/apps/commerce-avalara/commerce-avalara-lang/src/main/resources/content/Language.properties [hunk 2]
```diff
@@ -1,9 +1,21 @@
 account-number=Account Number
 avalara=Avalara
 avalara-tax-rate-description=Avalara Type Rate Description
-commerce-avalara-connecto-configuration-name=Commerce Avalara Connector Configuration Name
+avalara-tax-rates=Avalara Tax Rates
+avalara.tax.codes.commerce.health.status.description=Check if there are Avalara Tax Codes available to map to Tax Categories
```
Comment (dsitu) [inline modules/dxp/apps/commerce-avalara/commerce-avalara-lang/src/main/resources/content/Language.properties:5]: Should end with a period if its a sentence.
```suggestion
avalara.tax.codes.commerce.health.status.description=Check if there are Avalara tax codes available to map to tax categories.
```

File: modules/dxp/apps/commerce-avalara/commerce-avalara-connector/src/main/java/com/liferay/commerce/avalara/connector/internal/helper/CommerceAvalaraDispatchTriggerHelperImpl.java [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+		CommerceChannel commerceChannel = null;
+
+		try {
+			commerceChannel =
+				_commerceChannelLocalService.getCommerceChannelByGroupId(
+					commerceTaxMethod.getGroupId());
+		}
+		catch (PortalException portalException) {
+			_log.error("Could not get CommerceChannel", portalException);
```
Comment (dsitu) [inline modules/dxp/apps/commerce-avalara/commerce-avalara-connector/src/main/java/com/liferay/commerce/avalara/connector/internal/helper/CommerceAvalaraDispatchTriggerHelperImpl.java:197]:
```suggestion
			_log.error("Could not get commerce channel", portalException);
```

File: modules/dxp/apps/commerce-avalara/commerce-avalara-connector/src/main/java/com/liferay/commerce/avalara/connector/internal/helper/CommerceAvalaraConnectorHelperImpl.java [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+		String[] taxRatesByZipCodeLines = StringUtil.splitLines(
+			taxRateByZipCode);
+
+		CommerceCountry commerceCountryUS =
+			_commerceCountryLocalService.fetchCommerceCountry(
+				commerceTaxMethod.getCompanyId(), "US");
+
+		if (commerceCountryUS == null) {
+			commerceCountryUS = _addUnitedStates(commerceTaxMethod);
```
Comment (dsitu) [inline modules/dxp/apps/commerce-avalara/commerce-avalara-connector/src/main/java/com/liferay/commerce/avalara/connector/internal/helper/CommerceAvalaraConnectorHelperImpl.java:114]: Not sure if this is necessary, maybe just throw an exception?

File: modules/dxp/apps/commerce-avalara/commerce-avalara-connector-web/src/main/java/com/liferay/commerce/avalara/connector/web/internal/portlet/action/EditCommerceAvalaraConnectorMVCActionCommand.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Katie Nesterovich
+ */
+@Component(
+	enabled = false, immediate = true,
+	property = {
+		"javax.portlet.name=" + CommerceAvalaraPortletKeys.COMMERCE_AVALARA,
+		"mvc.command.name=/commerce_avalara_connector/edit_commerce_avalara_connector"
```
Comment (dsitu) [inline modules/dxp/apps/commerce-avalara/commerce-avalara-connector-web/src/main/java/com/liferay/commerce/avalara/connector/web/internal/portlet/action/EditCommerceAvalaraConnectorMVCActionCommand.java:46]: Lets name it based off `javax.portlet.name`
```suggestion
		"mvc.command.name=/commerce_avalara/edit_commerce_avalara_connector"
```

**Review by dsitu — CHANGES_REQUESTED** [review]

File: modules/dxp/apps/commerce-avalara/commerce-avalara-connector-web/src/main/resources/META-INF/resources/tabs/credentials.jspf [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+	cssClass="pt-4"
+>
+	<commerce-ui:panel>
+		<liferay-ui:error exception="<%= CommerceAvalaraConnectionException.class %>" message="the-connection-could-not-be-verified-because-the-provided-credentials-are-incorrect" />
+
+		<liferay-ui:success key="connectionSuccessful" message="connection-successful" />
+
+		<aui:form action="<%= editCommerceAvalaraCredentialsActionURL %>" method="post" name="fm">
+			<aui:input name="<%= Constants.CMD %>" type="hidden" value="<%= Constants.UPDATE %>" />
+			<aui:input id="avaglara-account-number" ignoreRequestValue="<%= true %>" label="account-number" name="accountNumber" value="<%= commerceAvalaraConnectorConfiguration.accountNumber() %>" />
```
[the remark this answers is not in the document]
Comment (katienesterovich) [reply modules/dxp/apps/commerce-avalara/commerce-avalara-connector-web/src/main/resources/META-INF/resources/tabs/credentials.jspf:33]: Good eye, thanks for catching!

File: modules/dxp/apps/commerce-avalara/commerce-avalara-connector/src/main/java/com/liferay/commerce/avalara/connector/internal/helper/CommerceAvalaraConnectorHelperImpl.java (hunk 4 above)
Comment (katienesterovich) [reply modules/dxp/apps/commerce-avalara/commerce-avalara-connector/src/main/java/com/liferay/commerce/avalara/connector/internal/helper/CommerceAvalaraConnectorHelperImpl.java:114]: In our setup guide we will advise the admin to run all the necessary health checks before they reach this stage, so ideally this check will never be true. However, we still want this process to execute if they haven't run the health checks (per earlier conversation with Corbin).

**Comment by katienesterovich** [comment]
@dsitu Hey Danny, thanks for a thorough review! 😄   I have added the edits you suggested, rebased to master, and re-ran the buildLang commit since it was causing merge conflicts.

**Comment by katienesterovich** [comment]
ci:test:sf

**Comment by katienesterovich** [comment]
ci:test:relevant

**Review by dsitu — CHANGES_REQUESTED** [review]

File: modules/dxp/apps/commerce-avalara/commerce-avalara-tax-engine-fixed-web/src/main/java/com/liferay/commerce/avalara/tax/engine/fixed/web/internal/frontend/constants/CommerceTaxRateSettingDataSetConstants.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ *
+ */
+
+package com.liferay.commerce.avalara.tax.engine.fixed.web.internal.frontend.constants;
+
+/**
+ * @author Alec Sloan
+ */
+public class CommerceTaxRateSettingDataSetConstants {
```
[the remark this answers is not in the document]
Comment (katienesterovich) [reply modules/dxp/apps/commerce-avalara/commerce-avalara-tax-engine-fixed-web/src/main/java/com/liferay/commerce/avalara/tax/engine/fixed/web/internal/frontend/constants/CommerceTaxRateSettingDataSetConstants.java:20]: Confirmed with Riccardo, moved to api.

**Comment by katienesterovich** [comment]
ci:test:sf

**Comment by katienesterovich** [comment]
ci:test:relevant

**Comment by katienesterovich** [comment]
@dsitu Thanks for the updated review! I have rebased and pushed suggested changes.

**Review by dsitu — APPROVED** [review]

**Comment by katienesterovich** [comment]
Rebased and resolved lang file conflict by re-running buildLang in a separate commit.

**Comment by katienesterovich** [comment]
Added c6af4eb, we overlooked the fact that Avalara provides rates in a decimal format (i.e .08 instead of 8). This commit converts the values. 

Note: as discussed with @riccardo-alberti, we may want to address this in commerce as a whole. The taxRate values were displayed properly (in Avalara Tax Rate tab), likely due to [PercentageFormatter](https://github.com/liferay/liferay-portal/blob/master/modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/percentage/PercentageFormatterImpl.java#L37), but the calculations were incorrect (1/100 of actual tax amount) as they were using the decimal format.

**Comment by katienesterovich** [comment]
Rebased and opened new PR https://github.com/liferay-commerce/liferay-portal/pull/886

**Pull request closed without merging**